### PR TITLE
Added OpenAM request logging to the VaultServiceActor. DSDEV-1837

### PR DESCRIPTION
**Pull request by kshakir**
Note to reviewer-- once this PR is approved, we need to updated both Chef and Jenkins before merging.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -28,3 +28,11 @@ libraryDependencies ++= {
 Revolver.settings
 
 javaOptions in Revolver.reStart += "-Dconfig.file=application.conf"
+
+// Copy over various properties
```
Comment (davidangb) [inline build.sbt:32]: I may be dense, but why is this necessary?

Comment (rushtong) [inline build.sbt:32]: I'm pretty sure that since restart fires up a forked jvm, it needs to know
about the config when doing so.

File: src/main/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectives.scala [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      // get the common name from the token
+      case token => commonNameFromToken(token)
+    }
+  }
+
+  // Partially based off DebuggingDirectives.logRequest, sans magnets, with requestInstance instead of mapRequest
+  // TODO: Any way to make this single log parameter implicit?
+  // The magnet pattern seems to always have at least one other parameter
+  // http://spray.io/blog/2012-12-13-the-magnet-pattern/
+  def logOpenAMRequest(log: LoggingAdapter) = commonNameFromCookie flatMap { commonName =>
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectives.scala]: this is the only difference between vault-api and vault-datamanagement. Any reason to keep the two in sync now? We'll certainly extract these into a common codebase soon; would it help to keep things in sync now?

(also thanks for the magnet pattern link. I just went on an educational roadtrip.)

File: src/test/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectivesSpec.scala [hunk 3]
```diff
@@ -0,0 +1,236 @@
+package org.broadinstitute.dsde.vault.common.openam
```
Comment (davidangb) [inline src/test/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectivesSpec.scala:1]: copy this file (or the relevant parts of it) back into vault-datamanagement?

File: src/test/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectivesSpec.scala [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+          fail("should not run this line")
+          completeOk
+        } ~> check {
+          rejection should be(MissingCookieRejection(OpenAMConfig.tokenCookie))
+        }
+      }
+
+    }
+
+    "when accessing OpenAM commonNameFromOptinalCookie directive" - {
```
Comment (davidangb) [inline src/test/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectivesSpec.scala]: typo: "optional"

File: src/test/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectivesSpec.scala [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        }
+      }
+
+    }
+
+    "when accessing OpenAM commonNameFromCookie directive" - {
+
+      "should get the common name" in {
+        Get("/hello") ~> OpenAMSession ~> commonNameFromCookie { commonName =>
+          commonName shouldNot be(empty)
```
Comment (davidangb) [inline src/test/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectivesSpec.scala:197]: we're modifying chef and jenkins anyway. Should we add the expected commonName value to the config files, just so we can test it? It's possibly overkill, but it gets us good test coverage on an important feature.

File: src/main/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectives.scala [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    requestInstance flatMap { request =>
+      val method = request.method.name
+      val uri = request.uri
+      var message = s"$commonName $method $uri"
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectives.scala]: commonName may or may not include a space, e.g. "David An" vs. "davidan". Having an unknown number of spaces makes parsing a space-delimited line difficult. Maybe make this tab-delimited? Or surround the commonName with some kind of markers?

File: src/main/scala/org/broadinstitute/dsde/vault/VaultServiceActor.scala [hunk 7]
```diff
@@ -33,11 +34,13 @@ class VaultServiceActor extends HttpServiceActor with ActorLogging {
 
   // this actor runs all routes
   def receive = runRoute(
-    uBAMIngest.routes ~ uBAMDescribe.routes ~ uBAMRedirect.routes ~
-    analysisIngest.routes ~ analysisDescribe.routes ~ analysisUpdate.routes ~ analysisRedirect.routes ~
-    lookupService.routes ~
-    swaggerService.routes ~ swaggerUiService
-    )
+    swaggerService.routes ~ swaggerUiService ~
+      logOpenAMRequest(log) {
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/vault/VaultServiceActor.scala]: I think we should use a different and dedicated LoggingAdapter here. That way, we can easily route audit logs to different appenders - we could have one file for audits and one for developer debug messages, and even a third for the combination of the two.

File: application.conf [hunk 8]
```diff
@@ -43,4 +43,12 @@ openam {
...
   testUserPassword=""
   tokenUrl=""
+  deploymentUri = "replace_with_openam_deployment_uri"
+  realm = "/apollo"
+  username = "replace_with_openam_username"
+  password = "replace_with_openam_password"
+  authIndex {
+    type = "module"
+    value = "DataStore"
+  }
```
Comment (davidangb) [inline application.conf:54]: I was stymied as to why my changes to logback.xml weren't affecting the logs you were writing. Turns out, I don't think our logging was _ever_ set up correctly. I'm not certain of the best way to do this, and would love to hear your opinion, but what I think we need is:

1) add akka-slf4j as a libraryDependency to build.sbt. Need to find the right version of this that matches our scala version.
2) add something like this to application.conf:

```
akka {
  loggers = ["akka.event.slf4j.Slf4jLogger"]
  loglevel = "DEBUG"
  logging-filter = "akka.event.slf4j.Slf4jLoggingFilter"
}
```

After all that, the akka system will honor logback.xml, and we should be good.

**Comment by davidangb** [comment]
Okay, first-pass review is done. Overall looks good, but some important comments inline.

The bulk of this code is a copy from vault-datamanagement. I checked for differences and didn't re-review the duplicate code.

File: build.sbt (hunk 1 above)
Comment (kshakir) [inline build.sbt:32]: Exactly, this helps with sbt revolver running in a separate VM. It effectively copies over the properties in .sbtopts.

Comment (davidangb) [inline build.sbt:32]: ahhhh, I wasn't thinking straight when I commented on this. I thought it was copying properties from application.conf. Now it makes sense; please ignore.

File: src/main/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectives.scala (hunk 2 above)
Comment (kshakir) [inline src/main/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectives.scala]: Will make this part of DSDEV-1820, creating a "common" repo. For now they will stay forked.

File: src/test/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectivesSpec.scala (hunk 3 above)
Comment (kshakir) [inline src/test/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectivesSpec.scala:1]: DSDEV-1820

File: src/test/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectivesSpec.scala (hunk 5 above)
Comment (kshakir) [inline src/test/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectivesSpec.scala:197]: Thanks! Fixed

File: src/main/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectives.scala (hunk 6 above)
Comment (kshakir) [inline src/main/scala/org/broadinstitute/dsde/vault/common/openam/OpenAMDirectives.scala]: Added [ELF](http://en.wikipedia.org/wiki/Extended_Log_Format)-ish double-quoted strings for the commonName.

File: src/main/scala/org/broadinstitute/dsde/vault/VaultServiceActor.scala (hunk 7 above)
Comment (kshakir) [inline src/main/scala/org/broadinstitute/dsde/vault/VaultServiceActor.scala]: Because the directive is outside an actor, I switched to using [slf4j directly](https://groups.google.com/forum/#!topic/akka-user/_bIiPKoGJXY). I set the "name" of the logger to "org.broadinstitute.dsde.vault.common.openam.OpenAMRequest", but I'm open to other names too.

File: application.conf (hunk 8 above)
Comment (kshakir) [inline application.conf:54]: Will create another ticket for this. From my understanding, we do need something like you describe above for logging to work.

A related issue: our unit tests based on the spray-testkit examples only access inner routes directly with a TestActor. We don't have integration tests of `VaultServiceActor.runRoute()`. So often during testing I would NOT see logs I expected since we never actually create the `VaultServiceActor`.

**Comment by kshakir** [comment]
Addressed comments. Ready for review of this code.

Before merging, once review here is done, still need to create the new tickets, update chef, update jenkins, and possibly squash some commits.

**Comment by davidangb** [comment]
Looks great to me. Back to you to handle squashing, chef, jenkins - and possibly conjur, if we want the target commonName to live there.

**Comment by kshakir** [comment]
IMHO the test username and password don't belong in chef, nor conjur-- only jenkins. I can see that currently the test user and password do exist in the two former locations, but I'm not clear on if/when/where this u/p are being used after being included by chef.

The chef generated application.conf files are for "main" use, not sbt testing, right? FYI, DM separates the two versions under `src/main/resources/application.conf` and `src/test/resources/application.conf`, while api just has the one version in the root.

File: application.conf [hunk 9]
```diff
@@ -43,4 +43,13 @@ openam {
   testUser=""
   testUserPassword=""
   tokenUrl=""
+  deploymentUri = "replace_with_openam_deployment_uri"
```
Comment (rpete) [inline application.conf:46]: What is this? Not sure what you mean by deployment URL?
